draft-ietf-jsonpath-base.md

```diff
@@ -480,32 +519,29 @@ operations which might be considered erroneous, such as indexing beyond the
 end of an array,
 simply result in fewer nodes being selected.
 
-But a selector doesn't just act on a single node: each selector acts on a
-list of nodes and produces a list of nodes, as follows.
-
-After the root selector, the remainder of the JSONPath is processed by passing
-lists of nodes from one selector to the next ending up with a list of nodes
-which is the result of
-applying the JSONPath to the input JSON value.
-
-Each selector acts on its input list of nodes as follows.
-For each node in
-the list, the selector selects zero or more nodes, each of which is a descendant
-of the node or the node itself.
-The output list of nodes of a selector is the concatenation of the lists
-of selected nodes for each input node.
-
-A specific, non-normative example will make this clearer.
-Suppose the input
-document is: `{"a":[{"b":0},{"b":1},{"c":2}]}`.
-As we will see later, the JSONPath `$.a[*].b` selects the following list of nodes: `0`, `1`.
+But a selector doesn't just act on a single node: a selector acts on
+each of the nodes in its input nodelist and selects an output
+nodelist for each of them, as follows, which are then combined using
+the combine1 algorithm to form the result nodelist of the selector.
+
+> Discussion: Do we allow selectors that have their own combine1
+> variant, or can we always use the same?
```

Comment:
Definitely worth discussion, but I think a common "combine" mechanism is simplest to understand.  Each selector defining its own forces the Path author to remember how each one behaves, which can get confusing fast.

Reply:
That is a good reason to stick with a common "combine1" model.

Reply:
Each selector can only select nodes from the subtree of the current node. So with each selector the number of potentially selectable nodes is reduced or stays constant at best. It can never increase ... so "combine" suggests "increase", is misleading and should either be "pick model" or "select model" or something similar.

Reply:
Well, a selector operates on the nodes in the input node list.
The "combine1" function is about combining the outputs that each such operation provides.

So if you have `[1], [2]` as the nodelist (identifying nodes by their values here), and the selector is `.[0]`, the output nodelists from each selector function invocation are `1` and `2` respectively.  combine1 would create an output nodelist `1, 2`.  This is a bit boring, but more complex selectors can leave a nodelist for each node in the input nodelist, so we need to discuss how to combine them:

`{a: {a: 1}},  {a: {a: 2}}` as input nodelist, `..a` as selector: The individual nodelists are  `{a: 1}, 1` and `{a: 2}, 2`.  combine1 turns this into, say, `{a: 1}, 1, {a: 2}, 2`.

Reply:
@goessner wrote:

> Oh well ... it's a simple "append" to the resulting nodelist. According to my mental model allowing only a single resulting nodelist, the sequence of ...
> 
> * `{a: 1}`
> * `1`
> * `{a: 2}`
> * `2`
> 
Just this. That corresponds to my "mental model", and directly in many implementations, some implementations.

> ... would be added, where order is implementation dependent of course (determinism).

Implementation dependent, yes, but I think the "natural order", would satisfy the minimally restrictive conditions identified by @glyn in #60.

Reply:
Yes, to spell this out, there is _one_ object with more than one key/value pairs and thus one source of non-determinism in this example (assuming the visitation ordering of `..` will be tied down in the spec). So the other possible resultant nodelist would  correspond to:
* `{a: 2}`
* `2`
* `{a: 1}`
* `1`

Reply:
This thread proposes an answer to the discussion point.
Maybe we can merge the PR first and then specifically address the discussion point.